File: src/main/java/com/ternbusty/takoyaki/console/InternalConsole.java
package com.ternbusty.takoyaki.console;

import com.ternbusty.takoyaki.ioloop.IoLoop;
import com.ternbusty.takoyaki.ipc.ScmRights;
import com.ternbusty.takoyaki.logger.Logger;
import com.ternbusty.takoyaki.syscall.Constants;
import com.ternbusty.takoyaki.syscall.Libc;
import com.ternbusty.takoyaki.syscall.PosixIO;

import java.io.IOException;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Internal PTY proxy for foreground {@code runc run} and {@code runc exec -t}
 * when no external {@code --console-socket} is given. The container side
 * opens the pty and sends the master fd back via SCM_RIGHTS; the I/O between
 * the master and the caller's stdin/stdout is then relayed by
 * {@link com.ternbusty.takoyaki.ioloop.Foreground}.
 *
 * <p>For run, a unix socket is bound at a temporary path in the bundle. For
 * exec, a pre-connected socketpair is used instead because the exec process
 * has already entered the container mount namespace (so a host path is
 * unreachable).
 *
 * <p>Nothing here waits on another thread. The runtime connects to the socket
 * itself before starting the init, and the init sends the master (or gives
 * up and closes its end) before it reports ready, so once create has
 * returned the master is either queued on the socket or never coming.
 */
public final class InternalConsole {
    private final String socketPath;
    private int listenFd = -1;
    private int masterFd = -1;

    private InternalConsole(String socketPath) {
        this.socketPath = socketPath;
    }

    /** Console socket path that should be passed to CreateCommand. */
    public String socketPath() { return socketPath; }

    /** The PTY master fd received via SCM_RIGHTS, or -1 if none was received. */
    public int masterFd() { return masterFd; }

    /**
     * Create the internal console socket for foreground {@code runc run}: a
     * non-blocking unix listener at a temporary path in the bundle. Call this
     * before the container init starts, then {@link #receiveMaster()} after
     * create has returned.
     *
     * @return the console, or null if the socket could not be set up
     */
    public static InternalConsole listenForRun(String bundlePath) {
        String path = bundlePath + "/internal-console.sock";
        // Remove stale socket from a previous run (unlink is idempotent).
        try { Files.deleteIfExists(Path.of(path)); } catch (IOException ignored) {}
        InternalConsole console = new InternalConsole(path);
        try (Arena arena = Arena.ofConfined()) {
            int fd = PosixIO.socket(Constants.AF_UNIX, Constants.SOCK_STREAM, 0);
            if (fd < 0) {
                Logger.warn("internal console: socket failed: " + Libc.strerror(Libc.errno()));
                return null;
            }
            console.listenFd = fd;
            if (PosixIO.bindUnix(arena, fd, path) < 0) {
                Logger.warn("internal console: bind " + path + " failed: " + Libc.strerror(Libc.errno()));
                console.stop();
                return null;
            }
            if (PosixIO.listen(fd, 1) < 0) {
                Logger.warn("internal console: listen failed: " + Libc.strerror(Libc.errno()));
                console.stop();
                return null;
            }
            IoLoop.setNonBlocking(fd);
        }
        Logger.debug("internal console: listening on " + path);
        return console;
    }

    /**
     * Take the PTY master fd the init sent, without blocking. Must be called
     * after create has returned; a missing connection or message means the
     * init could not set up a pty.
     *
     * @return the master fd, or -1
     */
    public int receiveMaster() {
        if (listenFd < 0) return -1;
        int connFd = PosixIO.accept(listenFd);
        PosixIO.close(listenFd);
        listenFd = -1;
        if (connFd < 0) {
            Logger.warn("internal console: no connection from init: " + Libc.strerror(Libc.errno()));
            return -1;
        }
        IoLoop.setNonBlocking(connFd);
        int fd = ScmRights.recvFd(connFd);
        PosixIO.close(connFd);
        if (fd < 0) {
            Logger.warn("internal console: init did not send a pty master");
            return -1;
        }
        clearONLCR(fd);
        Logger.debug("internal console: received master fd " + fd);
        masterFd = fd;
        return fd;
    }

    /**
     * Receive the PTY master fd from a pre-connected socketpair (exec path).
     * Blocks until the fd arrives or the peer closes; the caller must already
     * have closed its copy of the peer end.
     */
    public static int receiveMasterFromSocket(int sockFd) {
        int fd = ScmRights.recvFd(sockFd);
        if (fd >= 0) {
            clearONLCR(fd);
            Logger.debug("internal console (exec): received master fd " + fd);
        }
        return fd;
    }

    /**
     * Clear the ONLCR flag on a PTY master fd so that output newlines are
     * passed through verbatim instead of being converted to CR+LF. This
     * matches runc's {@code console.ClearONLCR()} and is essential for bats
     * test output comparisons.
     */
    public static void clearONLCR(int fd) {
        try (Arena arena = Arena.ofConfined()) {
            // struct termios (kernel version): c_iflag(4) c_oflag(4) c_cflag(4)
            // c_lflag(4) c_line(1) c_cc[19] = 36 bytes. Allocate 64 for safety.
            MemorySegment termios = arena.allocate(64);
            if (Libc.ioctl(fd, Constants.TCGETS, termios) != 0) {
                Logger.debug("clearONLCR: TCGETS failed: " + Libc.strerror(Libc.errno()));
                return;
            }
            int oflag = termios.get(ValueLayout.JAVA_INT, 4);
            oflag &= ~Constants.ONLCR;
            termios.set(ValueLayout.JAVA_INT, 4, oflag);
            if (Libc.ioctl(fd, Constants.TCSETS, termios) != 0) {
                Logger.debug("clearONLCR: TCSETS failed: " + Libc.strerror(Libc.errno()));
            }
        }
    }

    /** Clean up: close the listener and the master fd, remove the socket file. */
    public void stop() {
        if (listenFd >= 0) {
            PosixIO.close(listenFd);
            listenFd = -1;
        }
        if (masterFd >= 0) {
            PosixIO.close(masterFd);
            masterFd = -1;
        }
        try { Files.deleteIfExists(Path.of(socketPath)); } catch (IOException ignored) {}
    }
}

File: src/main/java/com/ternbusty/takoyaki/console/InternalConsole.kt
package com.ternbusty.takoyaki.console

import com.ternbusty.takoyaki.ipc.ScmRights
import com.ternbusty.takoyaki.logger.Logger
import com.ternbusty.takoyaki.syscall.Constants
import com.ternbusty.takoyaki.syscall.Libc
import com.ternbusty.takoyaki.syscall.PosixIO
import java.io.IOException
import java.lang.foreign.Arena
import java.lang.foreign.ValueLayout
import java.nio.file.Files
import java.nio.file.Path

/**
 * Internal PTY proxy for foreground `runc run` and `runc exec -t`
 * when no external `--console-socket` is given. Creates a temporary unix
 * socket, accepts one connection from the container init (which sends the PTY
 * master fd via SCM_RIGHTS), then copies I/O between the master and the
 * caller's stdin/stdout until the master closes (container exits).
 *
 * For exec, a pre-connected socketpair is used instead of a path-based
 * socket because the exec process has already entered the container mount
 * namespace (so a host path is unreachable).
 */
class InternalConsole private constructor(
    /** Console socket path that should be passed to CreateCommand. */
    val socketPath: String,
) {
    @Volatile
    private var masterFd: Int = -1

    @Volatile
    private var stopped: Boolean = false

    private var listenerThread: Thread? = null
    private var ioThread: Thread? = null

    companion object {
        /**
         * Create an internal console socket for foreground `runc run`. The
         * returned object owns a unix listener socket at a temporary path; call
         * [startListening] before the container init starts so the socket
         * is ready.
         */
        fun createForRun(bundlePath: String): InternalConsole {
            val path = "$bundlePath/internal-console.sock"
            // Remove stale socket from a previous run (unlink is idempotent).
            try {
                Files.deleteIfExists(Path.of(path))
            } catch (_: IOException) {
            }
            return InternalConsole(path)
        }

        /**
         * Receive the PTY master fd from a pre-connected socketpair (exec path).
         * Blocks until the fd arrives or the peer closes.
         */
        fun receiveMasterFromSocket(sockFd: Int): Int {
            val fd = ScmRights.recvFd(sockFd)
            if (fd >= 0) {
                clearONLCR(fd)
                Logger.debug("internal console (exec): received master fd $fd")
            }
            return fd
        }

        /**
         * Start I/O copying for a given master fd (used by both run and exec paths).
         */
        fun startIOCopyForFd(masterFd: Int): Thread {
            // master -> stdout thread (the important direction for bats tests).
            val reader = Thread({
                try {
                    Arena.ofConfined().use { arena ->
                        val buf = ByteArray(8192)
                        while (true) {
                            val n = PosixIO.read(arena, masterFd, buf)
                            if (n <= 0) break
                            System.out.write(buf, 0, n.toInt())
                            System.out.flush()
                        }
                    }
                } catch (_: Exception) {
                }
            }, "pty-to-stdout")
            reader.isDaemon = true
            reader.start()

            // stdin -> master thread (for interactive use).
            val writer = Thread({
                try {
                    Arena.ofConfined().use { arena ->
                        val buf = ByteArray(4096)
                        while (true) {
                            val n = System.`in`.read(buf)
                            if (n <= 0) break
                            val chunk = if (n == buf.size) buf else buf.copyOf(n)
                            PosixIO.write(arena, masterFd, chunk)
                        }
                    }
                } catch (_: Exception) {
                }
            }, "stdin-to-pty")
            writer.isDaemon = true
            writer.start()

            return reader
        }

        /**
         * Clear the ONLCR flag on a PTY master fd so that output newlines are
         * passed through verbatim instead of being converted to CR+LF. This
         * matches runc's `console.ClearONLCR()` and is essential for bats
         * test output comparisons.
         */
        fun clearONLCR(fd: Int) {
            Arena.ofConfined().use { arena ->
                // struct termios (kernel version): c_iflag(4) c_oflag(4) c_cflag(4)
                // c_lflag(4) c_line(1) c_cc[19] = 36 bytes. Allocate 64 for safety.
                val termios = arena.allocate(64)
                if (Libc.ioctl(fd, Constants.TCGETS, termios) != 0) {
                    Logger.debug("clearONLCR: TCGETS failed: ${Libc.strerror(Libc.errno())}")
                    return
                }
                var oflag = termios.get(ValueLayout.JAVA_INT, 4)
                oflag = oflag and Constants.ONLCR.inv()
                termios.set(ValueLayout.JAVA_INT, 4, oflag)
                if (Libc.ioctl(fd, Constants.TCSETS, termios) != 0) {
                    Logger.debug("clearONLCR: TCSETS failed: ${Libc.strerror(Libc.errno())}")
                }
            }
        }
    }

    /**
     * Start a thread that listens on the socket, accepts one connection, and
     * receives the PTY master fd via SCM_RIGHTS. The master fd is stashed for
     * [startIOCopy] to pick up.
     */
    fun startListening() {
        listenerThread = Thread({
            Arena.ofConfined().use { arena ->
                val listenFd = PosixIO.socket(Constants.AF_UNIX, Constants.SOCK_STREAM, 0)
                if (listenFd < 0) {
                    Logger.warn("internal console: socket failed: ${Libc.strerror(Libc.errno())}")
                    return@use
                }
                if (PosixIO.bindUnix(arena, listenFd, socketPath) < 0) {
                    Logger.warn(
                        "internal console: bind $socketPath failed: ${Libc.strerror(Libc.errno())}"
                    )
                    PosixIO.close(listenFd)
                    return@use
                }
                if (PosixIO.listen(listenFd, 1) < 0) {
                    Logger.warn("internal console: listen failed: ${Libc.strerror(Libc.errno())}")
                    PosixIO.close(listenFd)
                    return@use
                }
                Logger.debug("internal console: waiting for connection on $socketPath")
                val connFd = PosixIO.accept(listenFd)
                PosixIO.close(listenFd)
                if (connFd < 0) {
                    if (!stopped) {
                        Logger.warn(
                            "internal console: accept failed: ${Libc.strerror(Libc.errno())}"
                        )
                    }
                    return@use
                }
                masterFd = ScmRights.recvFd(connFd)
                PosixIO.close(connFd)
                if (masterFd >= 0) {
                    clearONLCR(masterFd)
                    Logger.debug("internal console: received master fd $masterFd")
                } else {
                    Logger.warn("internal console: failed to receive master fd")
                }
            }
        }, "internal-console-listener").also {
            it.isDaemon = true
            it.start()
        }
    }

    /** Wait for the listener thread to complete (connection established). */
    fun awaitMaster(timeoutMs: Long): Boolean {
        val thread = listenerThread ?: return false
        try {
            thread.join(timeoutMs)
        } catch (_: InterruptedException) {
        }
        return masterFd >= 0
    }

    /**
     * Start I/O copying between the PTY master and the caller's stdin/stdout.
     * Returns immediately; copying runs in background threads. Call
     * [stop] after the container exits to clean up.
     */
    fun startIOCopy() {
        if (masterFd < 0) return
        ioThread = startIOCopyForFd(masterFd)
    }

    /** Clean up: wait for I/O threads to drain, close master fd, remove socket file. */
    fun stop() {
        stopped = true
        // Let the reader thread drain remaining PTY data before closing the fd.
        // Without this, a race between waitForChild returning and the reader
        // reaching EOF can lose the last chunk of container output.
        ioThread?.let {
            try {
                it.join(5_000)
            } catch (_: InterruptedException) {
            }
        }
        if (masterFd >= 0) {
            PosixIO.close(masterFd)
            masterFd = -1
        }
        try {
            Files.deleteIfExists(Path.of(socketPath))
        } catch (_: IOException) {
        }
    }
}
